### PR TITLE
[CASSANDRA-19759] Validate table transactional modes when preparing transactions

### PR DESCRIPTION
**Pull request by stonewhitener**
CASSANDRA-19759

### CODE REVIEW

**Review by dcapwell — CHANGES_REQUESTED** [review]
Overall LGTM, but can you add tests in `org.apache.cassandra.cql3.statements.TransactionStatementTest`?

**Comment by stonewhitener** [comment]
@dcapwell Thank you for your review comment! I added three tests for the table validation introduced in this patch and modified the existing tests affected.

File: test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java [hunk 1]
```diff
@@ -33,14 +33,7 @@
...
 
-import static org.apache.cassandra.cql3.statements.TransactionStatement.DUPLICATE_TUPLE_NAME_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.EMPTY_TRANSACTION_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.ILLEGAL_RANGE_QUERY_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.NO_CONDITIONS_IN_UPDATES_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.NO_COUNTERS_IN_TXNS_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.NO_TIMESTAMPS_IN_UPDATES_MESSAGE;
-import static org.apache.cassandra.cql3.statements.TransactionStatement.SELECT_REFS_NEED_COLUMN_MESSAGE;
+import static org.apache.cassandra.cql3.statements.TransactionStatement.*;
```
Comment (dcapwell) [inline test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java]: please revert, this goes against our style guide; we don't like `.*`

**Review by dcapwell — CHANGES_REQUESTED** [review]

File: src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java [hunk 2]
```diff
@@ -100,6 +100,7 @@ public class TransactionStatement implements CQLStatement.CompositeCQLStatement,
     public static final String INCOMPLETE_PRIMARY_KEY_SELECT_MESSAGE = "SELECT must specify either all primary key elements or all partition key elements and LIMIT 1. In both cases partition key elements must be always specified with equality operators; %s %s";
     public static final String NO_CONDITIONS_IN_UPDATES_MESSAGE = "Updates within transactions may not specify their own conditions; %s statement %s";
     public static final String NO_TIMESTAMPS_IN_UPDATES_MESSAGE = "Updates within transactions may not specify custom timestamps; %s statement %s";
+    public static final String TRANSACTIONS_DISABLED_ON_TABLE_MESSAGE = "Accord transactions are disabled on a table (See transactional_mode in table option); %s statement %s";
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]:
```suggestion
    public static final String TRANSACTIONS_DISABLED_ON_TABLE_MESSAGE = "Accord transactions are disabled on table (See transactional_mode in table options); %s statement %s";
```

**Review by maedhroz — APPROVED** [review]
Left one minor nit, but otherwise LGTM

File: test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java (hunk 1 above)
Comment (stonewhitener) [reply test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java]: Done, but our `checkstyle.xml` does not include such a rule. Where can I find the evidence for your claim?

File: src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java (hunk 2 above)
Comment (stonewhitener) [reply src/java/org/apache/cassandra/cql3/statements/TransactionStatement.java]: Done! Thank you!

File: test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java (hunk 1 above)
Comment (dcapwell) [reply test/unit/org/apache/cassandra/cql3/statements/TransactionStatementTest.java]: https://cassandra.apache.org/_/development/code_style.html is what defines our style, but it overrides https://www.oracle.com/java/technologies/javase/codeconventions-fileorganization.htm

In both cases I don't see this documented... so I think we only approved moving away from `.*` in a mailing list thread and didn't update our guides?  Sorry about that =(

The only place that hits at that is `ide/idea/codeStyleSettings.xml` which makes it 100 elements before we replace with `.*`

**Review by dcapwell — APPROVED** [review]

**Comment by dcapwell** [comment]
merged https://github.com/apache/cassandra/commit/885f4362afdca43f0a98b03128b2b4993a6701e9

**Pull request closed without merging**